src/pykeen/utils.py

```diff
@@ -1447,6 +1447,77 @@ def ensure_complex(*xs: torch.Tensor) -> Iterable[torch.Tensor]:
         yield torch.view_as_complex(x)
 
 
+def iter_weisfeiler_lehman(
+    edge_index: torch.LongTensor,
+    max_iter: int,
+    num_nodes: Optional[int] = None,
+) -> Iterable[torch.Tensor]:
+    """
+    Iterate Weisfeiler-Lehman colors.
+
+    .. note::
+        the implementation creates intermediate dense tensors of shape `(num_nodes, num_colors)`
+
+    :param edge_index: shape: `(2, m)`
+        the edge list
+    :param max_iter:
+        the maximum number of iterations
+    :param num_nodes:
+        the number of nodes. If None, will be inferred from the edge index.
+
+    :raises ValueError:
+        if the number of nodes exceeds `torch.long` (this cannot happen in practice, as the edge index tensor
+        construction would already fail earlier)
+
+    :yields: the colors for each Weisfeiler-Lehman iteration
+    """
+    num_nodes = num_nodes or edge_index.max().item() + 1
+    colors = edge_index.new_zeros(size=(num_nodes,), dtype=torch.long)
```

Comment:
same device as edge index?

Reply:
Yes, that's why I use [`new_zeros`](https://pytorch.org/docs/stable/generated/torch.Tensor.new_zeros.html) 🙂